### PR TITLE
Migrate Config from dict to object

### PR DESCRIPTION
**Pull request by StoneMoe**
# Description
This PR:

- Fix https://github.com/bottlesdevs/Bottles/issues/2387
- Fix https://github.com/bottlesdevs/Bottles/issues/1692
- migrate bottle config from dict to object
- add a lot of type hints
- and many other bugfixs

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. 
Provide instructions so we can reproduce.
- [x] I tried my best to test it

### CODE REVIEW

File: bottles/backend/dlls/dll.py [hunk 1]
```diff
@@ -38,8 +40,9 @@ def __init__(self, version: str):
         self.base_path = self.get_base_path(version)
         self.check()
 
+    @staticmethod
```
Comment (StoneMoe) [inline bottles/backend/dlls/dll.py:43]: fix incorrect abstract method signature

File: bottles/backend/health.py [hunk 2]
```diff
@@ -18,9 +18,7 @@
 import os
 from bottles.backend.utils import yaml
 import shutil
-import platform
```
Comment (StoneMoe) [inline bottles/backend/health.py:21]: unused import

File: bottles/backend/health.py [hunk 3]
```diff
@@ -71,7 +69,7 @@ def __init__(self):
         }
         self.get_ram_data()
```
Comment (StoneMoe) [inline bottles/backend/health.py:71]: discouraged Python grammar

**Comment by StoneMoe** [comment]
I removed some unnecessary changes to reduce PR size.

**Comment by mirkobrombin** [comment]
There is a recursion when opening a bottle:
```
  File "/usr/lib/python3.10/dataclasses.py", line 1276, in _asdict_inner
    return type(obj)((_asdict_inner(k, dict_factory),
  File "/usr/lib/python3.10/dataclasses.py", line 1277, in <genexpr>
    _asdict_inner(v, dict_factory))
  File "/usr/lib/python3.10/dataclasses.py", line 1276, in _asdict_inner
    return type(obj)((_asdict_inner(k, dict_factory),
  File "/usr/lib/python3.10/dataclasses.py", line 1276, in <genexpr>
    return type(obj)((_asdict_inner(k, dict_factory),
  File "/usr/lib/python3.10/dataclasses.py", line 1243, in _asdict_inner
    if _is_dataclass_instance(obj):
  File "/usr/lib/python3.10/dataclasses.py", line 1208, in _is_dataclass_instance
    return hasattr(type(obj), _FIELDS)
RecursionError: maximum recursion depth exceeded while calling a Python object
```

**Comment by Kinsteen** [comment]
In frontend/utils/threading.py, L79, it should now be `RunAsync(func, *((None, True,)+args), **kwargs)`, as RunAsync has a new optional parameter. The comment can/should also be updated to include the `daemon` parameter that needs to be also populated with the default value

**Comment by StoneMoe** [comment]
> There is a recursion when opening a bottle:
> 
> ```
>   File "/usr/lib/python3.10/dataclasses.py", line 1276, in _asdict_inner
>     return type(obj)((_asdict_inner(k, dict_factory),
>   File "/usr/lib/python3.10/dataclasses.py", line 1277, in <genexpr>
>     _asdict_inner(v, dict_factory))
>   File "/usr/lib/python3.10/dataclasses.py", line 1276, in _asdict_inner
>     return type(obj)((_asdict_inner(k, dict_factory),
>   File "/usr/lib/python3.10/dataclasses.py", line 1276, in <genexpr>
>     return type(obj)((_asdict_inner(k, dict_factory),
>   File "/usr/lib/python3.10/dataclasses.py", line 1243, in _asdict_inner
>     if _is_dataclass_instance(obj):
>   File "/usr/lib/python3.10/dataclasses.py", line 1208, in _is_dataclass_instance
>     return hasattr(type(obj), _FIELDS)
> RecursionError: maximum recursion depth exceeded while calling a Python object
> ```

I can't reproduce this bug, may I have the bottle config yaml which triggered this issue?

File: bottles/backend/managers/manager.py [hunk 4]
```diff
@@ -856,22 +849,21 @@ def update_config(
             else:
                 config[key] = value
 
-        with open(os.path.join(bottle_path, "bottle.yml"), "w") as conf_file:
-            yaml.dump(config, conf_file, indent=4)
-            conf_file.close()
+        config.dump(os.path.join(bottle_path, "bottle.yml"))
 
-        config["Update_Date"] = str(datetime.now())
+        config.Update_Date = str(datetime.now())
```
[the remark this answers is not in the document]
Comment (StoneMoe) [reply bottles/backend/managers/manager.py:889]: it's introduced in 746081dfc20fc820ef3433d44586599058fb3fec, commit message says "Log syntax standardization." so I have no idea why it's like this

**Comment by Kinsteen** [comment]
This is out-of-scope, but I think that changing the variables names for DXVK, VKD3D etc with something more clear like dxvk_enabled and dxvk_version would improve a lot readability

**Comment by StoneMoe** [comment]
> This is out-of-scope, but I think that changing the variables names for DXVK, VKD3D etc with something more clear like dxvk_enabled and dxvk_version would improve a lot readability

I'm totally agree with that, but I'd prefer to keep it as is for now, because config migration needs more test and must be robust, which would be better to create a seperate merge request!

**Review by Kinsteen — APPROVED** [review]
Looks good to me, after the Discord comments are treated.

**Pull request merged**